src/context.rs

```diff
@@ -403,3 +403,107 @@ impl<A> FusedFuture for ReceiveFuture<A> {
         self.0.is_terminated()
     }
 }
+
+pub struct TickFuture<'a, A> {
+    state: TickState<'a, A>,
+}
+
+#[cfg(feature = "instrumentation")]
+impl<'a, A> TickFuture<'a, A> {
+    /// Return the handler's [`tracing::Span`](https://docs.rs/tracing/latest/tracing/struct.Span.html)
+    /// and the future, if the message is not a shut down. This can be used to log messages into the
+    /// span when required, such as if it is cancelled later due to a timeout.
+    pub fn with_span(self) -> (Option<tracing::Span>, TickFuture<'a, A>) {
+        let (state, span) = match self.state {
+            TickState::New { msg, actor, ctx } => {
+                let (fut, span) = TickState::running(msg, actor, ctx);
+                let state = TickState::Running {
+                    span: span.clone(),
+                    fut,
+                };
+
+                (state, span)
+            }
+            TickState::Running { span, fut } => {
+                let state = TickState::Running {
+                    span: span.clone(),
+                    fut,
+                };
+
+                (state, span)
+            }
+            TickState::Done => (TickState::Done, None),
+        };
+
+        let new = TickFuture { state };
+
+        (span.map(|span| span.0), new)
+    }
+}
+
+enum TickState<'a, A> {
+    New {
+        msg: Message<A>,
+        actor: &'a mut A,
+        ctx: &'a mut Context<A>,
+    },
+    Running {
+        fut: BoxFuture<'a, ControlFlow<()>>,
+        span: Option<HandlerSpan>,
+    },
+    Done,
+}
+
+impl<'a, A> TickState<'a, A> {
+    fn running(
+        msg: Message<A>,
+        actor: &'a mut A,
+        ctx: &'a mut Context<A>,
+    ) -> (BoxFuture<'a, ControlFlow<()>>, Option<HandlerSpan>) {
+        let (fut, span) = match msg.0 {
+            ActorMessage::ToOneActor(msg) => msg.handle(actor, ctx),
+            ActorMessage::ToAllActors(msg) => msg.handle(actor, ctx),
+            ActorMessage::Shutdown => {
+                return (
+                    Box::pin(async move {
+                        ctx.running = false;
+                        ControlFlow::Break(())
+                    }),
+                    None,
+                )
+            }
+        };
+
+        (fut, Some(span))
```

Comment:
If we would do this straight in the ctor, the span would be directly available and instead of the `with_span` function, we could have:

```rust
fn span(&self) -> Span
```

I'd make it panic if the future is complete already.

Is there a reason to access the span before polling the future? Otherwise we could also do it in the first poll and return `None` before or panic.

Thoughts?

Reply:
> > Is there a reason to access the span before polling the future? Otherwise we could also do it in the first poll and return None before or panic.
> 
> Do you mean to make the future resolve to `(Span, Fut)`?

No, I was thinking of having an _accessor_ to get the span out of the `Future`. In other words, instead of consuming self, it would just be &self. Something like:

```rust
match futures::select(sleep(1), ctx.tick(msg, &mut actor)) {
    Either::Left(((), unfinished_tick)) => {
        let _entered = unfinished_tick.span().enter();

        tracing::warn!(...);
    }
}
```

Reply:
This could be added in addition to the current function, I think. I don't think `fn span(&self) -> &Span` is possible, since `Shutdown` messages don't have a handle span, considering they don't invoke a handler, and are not currently instrumented.

Reply:
> This could be added in addition to the current function, I think.

Why not instead? I find the `with` API a bit awkward to use because we are not passing anything to it but it returns a tuple.

Reply:
If instead, it would need to be &mut self, unless it created the span upon future creation. The downside of that, though, is that it would start early.

The span creation is going to be leaked anyway by its nature, since when it starts is measurable.

Reply:
> If instead, it would need to be &mut self, unless it created the span upon future creation. The downside of that, though, is that it would start early.

We could do it on the first poll and have it return `None` before if we have to have `Option` in the signature anyway because of `Shutdown`.

Reply:
> We could do it on the first poll and have it return None before if we have to have Option in the signature anyway because of Shutdown.

This is essentially the behaviour now, but it you can create it early. Otherwise, you need to poll it and then still access the future which is usually quite unergonomic

Reply:
> > We could do it on the first poll and have it return None before if we have to have Option in the signature anyway because of Shutdown.
> 
> This is essentially the behaviour now, but it you can create it early. Otherwise, you need to poll it and then still access the future which is usually quite unergonomic

Is it? I am not sure I agree. Did you see my timeout example based on `select`?

I find it not clean that I have to create a `Span` outside of a code branch (before the poll), if it is only ever used within one code branch (the error case). As a rule of thumb, I tend to give variables a lifetime and scope that is as short/narrow as possible. 

Another view could be to say that `TickFuture` has a lifetime to `Context` anyway so you can't really do anything other than almost immediately poll it and thus, creating `Span` in the ctor doesn't falsify measurements much and always by the same amount for a given codebase whilst giving users the freedom to access the span whenever they want.

Reply:
> Is it? I am not sure I agree. Did you see my timeout example based on select?

I think this may be the exception rather than the rule, given that most methods of polling a future do not give access to the future after its first poll - it consumes it.

> Another view could be to say that TickFuture has a lifetime to Context anyway so you can't really do anything other than almost immediately poll it and thus, creating Span in the ctor doesn't falsify measurements much and always by the same amount for a given codebase whilst giving users the freedom to access the span whenever they want.

That's fair

Reply:
> > Is it? I am not sure I agree. Did you see my timeout example based on select?
> 
> I think this may be the exception rather than the rule, given that most methods of polling a future do not give access to the future after its first poll - it consumes it.

That is true too! I guess using `select` is more of a hack then?

What if we put the `Span` into a `RefCell<Option>`? We should be able to guarantee that no two mutable references exist at the same time so we should be able to create the span either on first poll or on first access through a `fn span(&self)` function.

Reply:
Another option would be to make it `get_or_create_span(&mut self)`?

Reply:
> Another option would be to make it `get_or_create_span(&mut self)`?

As long as there is no issue with pinning here, I think that works.

Should we go with `span` as function name and hide the "get or create" in the docs to make it less noisy?

Reply:
> As long as there is no issue with pinning here, I think that works.

The future is `Unpin`, so I think it should work?

> Should we go with span as function name and hide the "get or create" in the docs to make it less noisy?

I think `get_or_create` is better as it's more explicit and contains the explanation for why &mut self and not &self in its name, as well as it's intuitively showing the span would start from that call.

Reply:
Fine with me :)

Reply:
Done. I have a followup which fixes an issue where the Span would get created too early (I elaborated on this in xtra-community)